File: src/hotelier-core-app.Service/Implementation/PropertyService.cs
using AutoMapper;
using hotelier_core_app.Core.Constants;
using hotelier_core_app.Core.Helpers.Interface;
using hotelier_core_app.Domain.Commands.Interface;
using hotelier_core_app.Domain.Queries.Interface;
using hotelier_core_app.Model.DTOs.Request;
using hotelier_core_app.Model.DTOs.Response;
using hotelier_core_app.Model.Entities;
using hotelier_core_app.Service.Interface;
using Microsoft.AspNetCore.Identity;

namespace hotelier_core_app.Service.Implementation
{
    public class PropertyService : IPropertyService
    {
        private readonly IDBQueryRepository<Tenant> _tenantQueryRepository;
        private readonly UserManager<ApplicationUser> _userManager;
        private readonly IMapper _mapper;
        private readonly IDBCommandRepository<Address> _addressCommandRepository;
        private readonly IDBCommandRepository<Property> _propertyCommandRepository;
        private readonly IDBCommandRepository<AuditLog> _auditLogCommandRepository;
        private readonly IDBQueryRepository<Property> _propertyQueryRepository;
        private readonly IDBQueryRepository<Address> _addressQueryRepository;
        private readonly IUtility _utility;

        public PropertyService(IDBQueryRepository<Tenant> tenantQueryRepository,
            UserManager<ApplicationUser> userManager,
            IMapper mapper,
            IDBCommandRepository<Address> addressCommandRepository,
            IDBCommandRepository<Property> propertyCommandRepository,
            IDBCommandRepository<AuditLog> auditLogCommandRepository,
            IDBQueryRepository<Property> propertyQueryRepository,
            IDBQueryRepository<Address> addressQueryRepository,
            IUtility utility)
        {
            _tenantQueryRepository = tenantQueryRepository;
            _userManager = userManager;
            _mapper = mapper;
            _addressCommandRepository = addressCommandRepository;
            _propertyCommandRepository = propertyCommandRepository;
            _auditLogCommandRepository = auditLogCommandRepository;
            _propertyQueryRepository = propertyQueryRepository;
            _addressQueryRepository = addressQueryRepository;
            _utility = utility;
        }

        // add property
        public async Task<BaseResponse> AddProperty(AddPropertyRequestDTO request, AuditLog auditLog)
        {
            // validate input
            // verify that tenant exists
            // verify that user belongs to tenant
            // verify that user has permissions (via controller)

            var tenant = await _tenantQueryRepository.FindAsync(request.TenantId);
            if (tenant == null) return BaseResponse.Failure(ResponseMessages.TenantNotExisting);

            var currentUser = await _userManager.FindByEmailAsync(auditLog.PerformerEmail);
            if (currentUser == null) return BaseResponse.Failure(ResponseMessages.UserDoesNotExist);
            if (currentUser.TenantId != request.TenantId) return BaseResponse.Failure(ResponseMessages.UserNotInTenant);

            var address = _mapper.Map<Address>(request.Address);
            _addressCommandRepository.Add(address);

            var property = _mapper.Map<Property>(request);
            property.AddressId = address.Id;
            property.CreatedBy = auditLog.PerformedBy;
            property.CreationDate = DateTime.UtcNow;
            _propertyCommandRepository.Add(property);

            _auditLogCommandRepository.Add(auditLog);

            return BaseResponse.Success();
        }

        public async Task<BaseResponse> UpdateProperty(UpdatePropertyRequestDTO request, AuditLog auditLog)
        {
            var property = await _propertyQueryRepository.FindAsync(request.Id);
            if(property == null) return BaseResponse.Failure(ResponseMessages.PropertyNotFound);

            var currentUser = await _userManager.FindByEmailAsync(auditLog.PerformerEmail);
            if (currentUser == null) return BaseResponse.Failure(ResponseMessages.UserDoesNotExist);
            if (currentUser.TenantId != property.TenantId) return BaseResponse.Failure(ResponseMessages.UserNotInTenant);

            property.Name = request.Name;
            property.Description = request.Description;
            property.Image = request.Image;
            _propertyCommandRepository.Update(property);

            var address = _addressQueryRepository.Find(property.AddressId);
            address.Street = request.Address.Street;
            address.City = request.Address.City;
            address.State = request.Address.State;
            address.ZipCode = request.Address.ZipCode;
            address.Country = request.Address.Country;
            address.Latitude = request.Address.Latitude;
            address.Longitude = request.Address.Longitude;
            _addressCommandRepository.Update(address);

            return BaseResponse.Success();
        }

        public async Task<BaseResponse<PropertyResponseDTO>> GetById(long id)
        {
            var property = _propertyQueryRepository.FindByInclude(p => p.Id == id, p => p.Address);
            if (property == null) return BaseResponse<PropertyResponseDTO>.Failure(new PropertyResponseDTO(), ResponseMessages.PropertyNotFound);

            var response = _mapper.Map<PropertyResponseDTO>(property);
            return BaseResponse<PropertyResponseDTO>.Success(response);
        }

        public async Task<PageBaseResponse<List<PropertyResponseDTO>>> GetTenantPropertyList(GetPropertiesInputDTO input)
        {
            var properties = await _propertyQueryRepository.GetByAsync(p => p.TenantId == input.TenantId);
            var paginated = _utility.Paginate(properties, input.PageNumber, input.PageSize);
            var response = _mapper.Map<List<PropertyResponseDTO>>(paginated);

            return PageBaseResponse<List<PropertyResponseDTO>>.Success(response, count: response.Count(), totalPageCount: properties.Count(), pageSize: input.PageSize, pageNumber: input.PageNumber);
        }
    }
}

File: hotel-ms-ui/src/services/property.service.ts
import api from '../lib/axios';
import type { Property, ApiResponse, PageApiResponse } from '../types';

function mapProperty(r: {
  id: number;
  name: string;
  description?: string;
  image?: string;
  creationDate?: string;
  lastModifiedDate?: string;
  address?: {
    street?: string;
    city?: string;
    state?: string;
    zipCode?: string;
    country?: string;
    latitude?: number;
    longitude?: number;
  };
}): Property {
  return {
    ...r,
    city: r.address?.city,
    country: r.address?.country,
    createdAt: r.creationDate ?? new Date().toISOString(),
  } as Property;
}

export const propertyService = {
  /**
   * GET /api/v1/Properties?tenantId=&name=&pageNumber=&pageSize=
   */
  async getAll(params?: {
    tenantId?: number;
    name?: string;
    pageNumber?: number;
    pageSize?: number;
  }): Promise<Property[]> {
    const response = await api.get<PageApiResponse<unknown[]>>('/api/v1/Properties', {
      params: { pageSize: 100, ...params },
    });
    const raw = response.data?.data;
    if (!Array.isArray(raw)) return [];
    return raw.map((r) => mapProperty(r as Parameters<typeof mapProperty>[0]));
  },

  /**
   * GET /api/v1/Properties/{id}
   */
  async getById(id: number): Promise<Property> {
    const response = await api.get<ApiResponse<unknown>>(`/api/v1/Properties/${id}`);
    return mapProperty(response.data.data as Parameters<typeof mapProperty>[0]);
  },

  /**
   * POST /api/v1/Properties
   * Body: { name, description, image, tenantId, address: { street, city, state, zipCode, country } }
   */
  /**
   * POST /api/v1/Properties
   * Backend returns BaseResponse (no data). Caller must reload after success.
   */
  async create(request: {
    name: string;
    description: string;
    image: string;
    tenantId: number;
    address: { street?: string; city?: string; state?: string; zipCode?: string; country?: string; latitude?: number; longitude?: number };
  }): Promise<void> {
    await api.post('/api/v1/Properties', request);
  },

  /**
   * PUT /api/v1/Properties
   * Backend returns BaseResponse (no data). Caller must reload after success.
   */
  async update(request: {
    id: number;
    name: string;
    description: string;
    image: string;
    address: { street?: string; city?: string; state?: string; zipCode?: string; country?: string; latitude?: number; longitude?: number };
  }): Promise<void> {
    await api.put('/api/v1/Properties', request);
  },
};
